### ISSUE TITLE
🔬 Support for calibration routines

### ISSUE THREAD

**Issue by echavarria-mqv** [Munich-Quantum-Software-Stack/QDMI#108]
The `v0.1` version of QDMI had dedicated routines for calibration
```c++
int QDMI_device_quality_check(QDMI_Device dev, double *result);
int QDMI_device_quality_limit(QDMI_Device dev, double *result);
int QDMI_device_quality_calibrate(QDMI_Device dev);
```

These have been purposely removed as part of the transition towards `v1` due to their limited usability.
However, calibration is an important aspect of integrating QC systems into HPC environments.
Hence, QDMI should allow to trigger calibration for devices.

@martin-knudsen: Just putting in motion the (re-)integration of this feature as part of the delayed deliverable on "continuous re-calibration" for MUNIQC-SC.

**Comment by burgholzer** [comment]
Thanks for creating this issue. I updated the title as well as the description of the issue to provide a little more information.

Adding calibration functionality to QDMI definitely makes sense.
It is just a matter of finding the right way to incorporate the required workflows.
Let's use this issue for brainstorming possible solutions.

@ystade and I briefly discussed this some time ago and we came to the conclusion that the current structure of QDMI would already allow one to incorporate calibration. The proposal on the table would be to

- Define a new [`QDMI_PROGRAM_FORMAT_T`](https://munich-quantum-software-stack.github.io/QDMI/enums_8h.html#a3eff3252203fcad146d5896e4466ec09) for triggering calibration, e.g., `QDMI_PROGRAM_FORMAT_CALIBRATION`, and use the job submission system to trigger a "calibration job". In the end, it _is_ just another job to be run on the system.

- I'd assume there needs to be some kind of way to ask the device whether it needs to be calibrated, right? For that, the proposal would be to define a new [`QDMI_DEVICE_PROPERTY_T`](https://munich-quantum-software-stack.github.io/QDMI/enums_8h.html#ad251d8ae8fbbe9a5c7a10d66b243d526) that allows to check exactly that, e.g., `QDMI_DEVICE_PROPERTY_NEEDSCALIBRATION"

Two questions that are less clear to me:
- Is there any need for further properties/queries/results?
- Is this design, where the client calls the respective functions, fine or is this something that we want the driver to manage? In the latter case, we would probably need to define new functions as part of the driver and the device interface, which are not exposed to the client.

**Comment by echavarria-mqv** [comment]
Sounds cool. We should keep in mind that quantum devices tend to have multiple calibration routines though.

1. How would you suggest to trigger a specific calibration? 
2. I guess we would also need to be able to query how many different types of calibration can be triggered.
3. Providers should be allowed to decide if calibration can be triggered from the QDMI client side. I guess we need something like `QDMI_DEVICE_PROPERTY_CALIBRATIONSUPPORT`.

**Comment by burgholzer** [comment]
> Sounds cool. We should keep in mind that quantum devices tend to have multiple calibration routines though.
> 
> 1. How would you suggest to trigger a specific calibration?
> 2. I guess we would also need to be able to query how many different types of calibration can be triggered.
> 3. Providers should be allowed to decide if calibration can be triggered from the QDMI client side. I guess we need something like `QDMI_DEVICE_PROPERTY_CALIBRATIONSUPPORT`.

Thanks for the comments!
I think the third part is the easiest to address, as it is currently supported. The job creation routine will report a "not supported" error, when a particular program format is not supported by the device.

The other two are a bit harder to address. At first, it seems hard to me to unify the different kinds of calibration across vendors and/or technologies. Standardization is probably hard here.
Given how we already have the capability to set parameters on a job, we could probably solve the problem of configuring the calibration that way, i.e., one would first create a job with a "calibration" program format and subsequently set some respective parameters on that job, before eventually submitting the job.
At first, while the precise parameters might be uncertain, one could make use of the custom parameters.

Does that make sense?
@ystade a sanity check from your side would probably also help here.

**Comment by ystade** [comment]
> @ystade a sanity check from your side would probably also help here.

@burgholzer I actually wanted to reply something along the same lines as you did above. Hence, I am fully in line with your view. Especially, I think as the way of calibration is not standardized at this point in time, the implementation via dedicated calibration functions is even more difficult because my feeling suggests that this requires more standardization as the aforementioned proposal.

I just want to point out one more difficulty which I even do not know whether it is relevant. I think one part of point 3. of @echavarria-lrz reply

> 3\. Providers should be allowed to decide if calibration can be triggered from the QDMI client side. I guess we need something like `QDMI_DEVICE_PROPERTY_CALIBRATIONSUPPORT`.

is that the client might set a limit fidelity and whenever the device falls below this fidelity limit it automatically starts a calibration round. In this case, the client does not submit a job to the device as the calibration is started by the device itself. However, once #117 is solved, such a limit can be set via a device parameter.

**Comment by echavarria-mqv** [comment]
I don't think this is a feature we need to prioritize at the moment (or even support at any point):

> the client might set a limit fidelity

It could be effectively managed by setting a limit on the client side and triggering a calibration round using the appropriate routine.

Regarding the following comment:

> the device falls below this fidelity limit it automatically starts a calibration round

I wouldn't advise automating this process with such a straightforward approach (on any side). At the LRZ, we have encountered cases where devices became unstable precisely because of this. The solution we will adopt at the LRZ, for example, is to schedule calibration on the client side (whenever supported by the backend) based on, for example, the current load, time of the day, telemetry data, and so on.

**Comment by burgholzer** [comment]
Alright. Thanks for the productive discussion here. It seems that we have reached a state of agreement when it comes to the initial support for calibration routines.
Thus, I would propose to go forward with implementing the proposal in https://github.com/Munich-Quantum-Software-Stack/QDMI/issues/108#issuecomment-2520238480.
It might be worth to use that opportunity to also add a small bit of documentation on this in addition to the usual API docs.
